### PR TITLE
Do not optimize common block variable that is potentially aliased

### PR DESCRIPTION
**Pull request by yaoliu-hnc**
Fix the bug while using common in module and subroutine, the testcase is following:
```fortran
module mod1
  integer :: mi1
  common /com1/ mi1
contains
  subroutine newmi1()
    mi1 = 30
  end subroutine newmi1
end module mod1

subroutine sub(x)
  use mod1
  integer :: si1, x
  common /com1/ si1
  si1 = 40
  mi1 = 50
  if(si1 == mi1)  x = 1
end subroutine sub  

program p
  use mod1
  integer :: res(5)
  integer :: pi1
  common /com1/ pi1
  res = 0
  pi1 = 10
  mi1 = 20

  if(pi1 == mi1)    res(1) = 1
  if(pi1 == 20)     res(2) = 1
  call newmi1()
  mi1 = 20
  pi1 = 10
  if(pi1 == mi1)    res(3) = 1
  if(mi1 == 10)     res(4) = 1
  call sub(res(5))

  do i = 1,5
    if (res(i) == 1) print *, "case ", i, "passed."
    if (res(i) == 0) print *, "case ", i, "failed."
  end do
end program p                               
```
the result while -O0 is correct:
```
 case             1 passed.
 case             2 passed.
 case             3 passed.
 case             4 passed.
 case             5 passed.
```
While using `-O1` option, the unexpected result appears as:
```
 case             1 failed.
 case             2 passed.
 case             3 failed.
 case             4 passed.
 case             5 failed.
```
Then, using `-O2`, `-O3` and `-Os`, there is 2 more error:
```
 case             1 failed.
 case             2 failed.
 case             3 failed.
 case             4 failed.
 case             5 failed.
 ```

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
Not clear what the issues that is being fixed here is. Could you explain in the description?

**Comment by yaoliu-hnc** [comment]
> Not clear what the issues that is being fixed here is. Could you explain in the description?

OK, Thank you.

**Comment by yaoliu-hnc** [comment]
I am working for a better solution. Change it to a draft PR.

**Comment by yaoliu-hnc** [comment]
This bugfix has been rebuild. Code modified in a new commit.

**Comment by yaoliu-hnc** [comment]
@kiranchandramohan We reopened this PR. Bugfix has been rebuild. Please have a look.

**Comment by yaoliu-hnc** [comment]
@kiranchandramohan Hi, would you please have a look? 
The CI fails at the stage of downloading artifacts. This happened before in another PR(#1038) but it went away shortly afterwards.
What should we do in this situation? Thanks.

**Comment by bryanpkc** [comment]
@yaoliu-hnc  Try modifying the branch, then force-pushing the original version of the branch. This should re-trigger the GitHub CI workflow.

**Comment by yaoliu-hnc** [comment]
> @yaoliu-hnc Try modifying the branch, then force-pushing the original version of the branch. This should re-trigger the GitHub CI workflow.

OK. Thank you :)

**Comment by yaoliu-hnc** [comment]
@kiranchandramohan The code for checking if two common variables are interfering each other are refactored.
CI fails at the stage of building libpgmath. I see that PR(#1060)and PR(#1058)have the same problem at the moment. Should I modify and force-push again?

File: tools/shared/nmeutil.c [hunk 1]
```diff
@@ -1245,6 +1245,12 @@ conflict(int nm1, int nm2)
             if (sptr2 == SOC_SPTR(t1))
               return CONFLICT;
       }
+
+      /* If sptr1 and sptr2 are overlapping common block variable,
+       * reference handing way of equivalence statement, do not optimize.
```
Comment (bryanpkc) [inline tools/shared/nmeutil.c:1250]: The comment is hard to understand. I think you mean:
```c
/* If sptr1 and sptr2 are overlapping common block variables, do not
 * optimize their loads/stores, like how variables in an equivalence
 * statement are handled.
 */
```

File: tools/flang1/flang1exe/optutil.c [hunk 2]
```diff
@@ -802,6 +802,16 @@ def_ok(int def, int fgx, int stmt)
     return FALSE;
   }
 
+  /* If sym is common block variable, reference handing way of
+   * equivalence statement, do not optimize.
+   */
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/optutil.c:807]: This comment is hard to understand. I suggest the following instead:
```c
/* If sym is a common block variable (and potentially aliased), do not propagate
 * its definition, like how variables in an equivalent statement are handled.
 */
```

File: test/f90_correct/src/common_01.f90 [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  integer, parameter :: n = 6
+  integer :: res(n), rres(n)
+  integer :: pi1
+  common /com1/ pi1
+
+  res = 0
+  rres= 1
+  pi1 = 10
+  mi1 = 20
+  if (pi1 == mi1) res(1) = 1
```
Comment (bryanpkc) [inline test/f90_correct/src/common_01.f90:28]: Not a big deal, but if you use the `check_mod` module, you could declare the `reslt` and `expct` arrays as `logical(n)` and these `if` statements could be reduced to:
```fortran
reslt(1:3) = (/pi1 == mi1, pi1 == 20, mi1 == 20/)
```

Comment (qiaozhang-hnc) [reply test/f90_correct/src/common_01.f90]: I've modified the code as suggested. Thanks!

Comment (yaoliu-hnc) [reply test/f90_correct/src/common_01.f90]: I've modified the code as suggested. Thanks!

**Review by bryanpkc — APPROVED** [review]
LGTM.

**Review by pawosm-arm — APPROVED** [review]

**Review by shivaramaarao — APPROVED** [review]
LGTM

**Pull request merged**